app/controllers/orders_controller.rb

```diff
@@ -0,0 +1,113 @@
+class OrdersController < ApplicationController
+
+  before_action :find_order, only: [:update, :edit, :show, :my_order, :change_status]
+
+  before_action :find_cart_by_session, only: [:new]
+  before_action :find_cart_by_order, only: [:update, :show]
+
+
+  def new
+    @order = Order.new
+  end
+
+  def create
+    @order = Order.new(order_params)
+    @order.cart_id = session[:cart_id]
+
+    if @order.save
+      flash[:status] = :success
+      redirect_to edit_order_path(@order)
+    else
+      flash.now[:status] = :failure
+      flash.now[:result_text] = "Could not place your order"
+      flash.now[:messages] = @order.errors.messages
+
+      render :new, status: :bad_request
+    end
+  end
+
+  def show; end
+
+  def edit; end
+
+  def update
+    @order.assign_attributes(order_params)
+
+    if @order.save
+      @cart.cartitems.each do |cartitem|
+        product = Product.find(cartitem.product_id)
+        product.stock = product.new_stock(cartitem.quantity)
+        product.visible = false if product.stock == 0
+        unless product.save
+          flash.now[:status] = :failure
+          flash.now[:result_text] = "Our #{product.name} is out of stock, please update your cart!"
+          redirect_to cart_path(@cart)
+        end
+      end
+
+      flash[:status] = :success
+      flash[:result_text] = "Your order has been placed !"
+      flash[:order_number] = "Order number: #{@order.id}"
+      session[:cart_id] = nil
+      redirect_to order_path(@order)
+    else
+
+      flash.now[:status] = :failure
+      flash.now[:result_text] = "Could not place your order"
+      flash.now[:messages] = @order.errors.messages
```

Comment:
There are many places in this app where you have code similar to this, setting the status to `:failure` and adding `:result_text` and `:messages`. Could you make this a helper method in the `ApplicationController`, something along the lines of:

```ruby
def report_failure(result_text, model=nil, current_request: true)
  target = current_request ? flash.now : flash
  target[:status] = :failure
  target[:result_text] = result_text
  if model
    target[:messages] = model.errors.messages
  end
end
```

Then here you could call it with `report_failure("Could not place your order", @order)`

Reply:
That's a great idea!